### ISSUE TITLE
pubspec.lock is ignored when upgrading to flutter 3.7.0

### ISSUE THREAD

**Issue by Jordan-Nelson** [flutter/flutter#119091]
## Steps to Reproduce

1. Install flutter 3.3.10
2. Create a project that has some dependencies which are pinned to a version that is not the latest. For example, if package foo has version 1.0.1 as the latest version, lock on 1.0.0.

pubspec.yaml
```yaml
dependencies:
  flutter:
    sdk: flutter
  foo: ^1.0.0
```
pubspec.lock
```
# Generated by pub
# See https://dart.dev/tools/pub/glossary#lockfile
packages:
  foo:
    dependency: "direct main"
    description:
      name: foo
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.0.0"
```
3. Run `flutter pub get` and notice that the package version is NOT upgraded since there is a lock file.
3. Upgrade to flutter 3.7.0
4. Run  `flutter pub get` and notice that the package version IS upgraded because the lock file was totally regenerated and the old one was ignored.

**Expected results:** 

pubspec.lock should always be honored. 

**Actual results:** 

pubspec.lock is ignored when upgrading to flutter 3.7.0

**Comment by exaby73** [comment]
Hello @Jordan-Nelson. Following your steps, with the `go_router` package instead of `foo` since it's not null safe, I cannot reproduce the issue.

I generated a project with 3.3.10, ran `flutter pub add go_router:6.0.0` (latest version is 6.0.1). I then changed the version in `pubspec.yaml` from `go_router: 6.0.0` to `go_router: ^6.0.0`. Running `flutter pub get` in both verisons, 3.3.10 and 3.7.0 produces the same result in the lock file. Here's a snippet after running `flutter pub get` on 3.7.0:

```yaml
...
  go_router:
    dependency: "direct main"
    description:
      name: go_router
      sha256: "788a89932142b0059c319e3153e73308e0fc8bbab78e9a66b0128423090c4e01"
      url: "https://pub.dev"
    source: hosted
    version: "6.0.0"
...
```

**Comment by Jordan-Nelson** [comment]
@exaby73 - That isn't consistent with what I see. That lock file you shared contains commit shas, which I only see when running `flutter pub get` with v3.7.0. Are you sure the lock file was initially created with 3.3.10?

I have created a public repo to help reproduce this. This project has one dependency, go_router, which is set to `go_router: ^6.0.0` in pubspec.yaml, and it is locked to `"6.0.0"` in the lockfile.

Steps to reproduce:
1. Use flutter 3.7.0
2. Run `git clone https://github.com/Jordan-Nelson/my_app.git`
3. Run `cd my_app && flutter pub get`
4. Observe the log output, which includes `go_router 6.0.1 (was 6.0.0)` indicating go_router was upgraded and the lock file was ignored.
5. Run `git diff | grep go_router -A 6` to print out the portion of the diff showing go_router has changes in the lock file.

Here is a sample of the log output when I run these exact steps.

```
nejrd@147dda5df20c tmp % git clone https://github.com/Jordan-Nelson/my_app.git
Cloning into 'my_app'...
remote: Enumerating objects: 180, done.
remote: Counting objects: 100% (180/180), done.
remote: Compressing objects: 100% (126/126), done.
remote: Total 180 (delta 25), reused 180 (delta 25), pack-reused 0
Receiving objects: 100% (180/180), 279.43 KiB | 1.68 MiB/s, done.
Resolving deltas: 100% (25/25), done.
nejrd@147dda5df20c tmp % cd my_app 
nejrd@147dda5df20c my_app % flutter pub get
Running "flutter pub get" in my_app...
Resolving dependencies... (1.1s)
* async 2.10.0 (was 2.9.0)
* boolean_selector 2.1.1 (was 2.1.0)
* characters 1.2.1 (was 1.2.1)
* clock 1.1.1 (was 1.1.1)
* collection 1.17.0 (was 1.16.0)
* cupertino_icons 1.0.5 (was 1.0.5)
* fake_async 1.3.1 (was 1.3.1)
* flutter_lints 2.0.1 (was 2.0.1)
* go_router 6.0.1 (was 6.0.0)
* js 0.6.5 (was 0.6.4) (0.6.6 available)
* lints 2.0.1 (was 2.0.1)
* logging 1.1.0 (was 1.1.0)
* matcher 0.12.13 (was 0.12.12) (0.12.14 available)
* material_color_utilities 0.2.0 (was 0.1.5)
* meta 1.8.0 (was 1.8.0)
* path 1.8.2 (was 1.8.2) (1.8.3 available)
* source_span 1.9.1 (was 1.9.0)
* stack_trace 1.11.0 (was 1.10.0)
* stream_channel 2.1.1 (was 2.1.0)
* string_scanner 1.2.0 (was 1.1.1)
* term_glyph 1.2.1 (was 1.2.1)
* test_api 0.4.16 (was 0.4.12) (0.4.18 available)
* vector_math 2.1.4 (was 2.1.2)
Changed 23 dependencies!
nejrd@147dda5df20c my_app % git diff | grep go_router -A 6
       name: go_router
-      url: "https://pub.dartlang.org"
+      sha256: f611d4396469c46db1c61e934a86e2a590ce02de2a6050d01f677879ce151f4a
+      url: "https://pub.dev"
     source: hosted
-    version: "6.0.0"
+    version: "6.0.1"
```

**Comment by Jordan-Nelson** [comment]
It looks like Dart 2.19 requires regenerating the lock file as there was a change to switch to commit shas. No issue with that. However, I am not sure why regenerating the lock file would ignore the current one. It could regenerate the lock file while still honoring the versions in the existing lock file.

**Comment by Jordan-Nelson** [comment]
For some context, I first saw this when a CI pipeline broke. Because the lock file was ignored, the latest version of a few dependencies were brought in, some of which had breaking changes.

**Comment by exaby73** [comment]
## Triage report

Removing the package and re-adding it reproduced this for me. The lock file seems to be regenerated, ignoring the dependency constraints in the previous version.

### Steps to reproduce

- Create a new Flutter project with `3.3.10`. Easiest way to manage multiple versions of flutter is using [FVM](https://fvm.app)
- Add a package that is one minor version behind. eg: `go_router: 6.0.0`. Latest version is `6.0.1` as of writing this
- Run `flutter pub get`. The generated lock file should have the dependency locked to `6.0.0`.
- Change the version from `go_router: 6.0.0` to `go_router: ^6.0.0`. Running `flutter pub get` now should yield the same results. The package should still be locked to `6.0.0` unless you run `flutter pub upgrade`
- Switch the flutter version to `3.7.0`
- Run `flutter pub get` again. Notice that the lock file now has the `sha256` and the version was upgraded to `6.0.1` despite the previous version of the lock file having specified `6.0.0`

<details><summary><code>flutter doctor -v</code> (Master)</summary>

```
[!] Flutter (Channel master, 3.7.0-27.0.pre.25, on macOS 13.1 22C65 darwin-arm64, locale en-US)
    • Flutter version 3.7.0-27.0.pre.25 on channel master at /Users/nabeelparkar/fvm/versions/master
    ! Warning: `dart` on your path resolves to /opt/homebrew/Cellar/dart/2.18.6/libexec/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nabeelparkar/fvm/versions/master. Consider adding /Users/nabeelparkar/fvm/versions/master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision c9affdba9d (67 minutes ago), 2023-01-25 21:25:02 -0800
    • Engine revision 8f1e5dc1b1
    • Dart version 3.0.0 (build 3.0.0-157.0.dev)
    • DevTools version 2.20.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nabeelparkar/Library/Android/sdk/
    • Platform android-33, build-tools 33.0.0
    • ANDROID_SDK_ROOT = /Users/nabeelparkar/Library/Android/sdk/
    • Java binary at: /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9123335/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /Applications/Brave Browser.app/Contents/MacOS/Brave Browser

[✓] Android Studio (version 2021.3)
    • Android Studio at /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9123335/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] IntelliJ IDEA Ultimate Edition (version 2022.3.1)
    • IntelliJ at /Users/nabeelparkar/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 71.3.6
    • Dart plugin version 223.8214.16

[✓] IntelliJ IDEA Ultimate Edition (version 2022.3.1)
    • IntelliJ at /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/223.8214.52/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.74.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.56.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.1 22C65 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Brave Browser 109.1.47.171

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```

</details>

**Comment by stuartmorgan** [comment]
@mit-mit Was there any intentional `pub` behavior change around unlocking?

**Comment by mit-mit** [comment]
Not to my knowledge, but maybe @jonasfj or @sigurdm can comment?

**Comment by sigurdm** [comment]
This change of behavior is unintentional :(

I can reproduce, and will try to investigate why this happens.

**Comment by sigurdm** [comment]
Ah - now I realize the bug here.

We switched from pub.dartlang.org to pub.dev as the default hosted source in 2.19. That makes pub see the packages as different (because they now come from a different host) and redoes the resolution.

**Comment by sigurdm** [comment]
Not sure what is the best fix, given that this already is out

**Comment by sigurdm** [comment]
So after looking at the code, I see that we are normalizing the two different hosts in most places, but missed when reading from the lockfile.

I'm working on a fix, and after that a hotfix for the 2.19 series. Apologies for this regression

**Comment by sigurdm** [comment]
Hotfix releases Dart 2.9.1 and Flutter 3.7.1 has landed with a fix.
